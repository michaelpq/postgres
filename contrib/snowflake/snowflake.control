# snowflake extension
comment = 'snowflake - sequence access method'
default_version = '1.0'
module_pathname = '$libdir/snowflake'
relocatable = true
