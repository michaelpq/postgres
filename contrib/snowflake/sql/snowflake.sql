CREATE EXTENSION snowflake;

CREATE SEQUENCE snowflake_seq USING snowflake;

SET snowflake.machine_id = 2000; -- error
SET snowflake.machine_id = 4; -- ok
SELECT machine, counter FROM snowflake_get(nextval('snowflake_seq'));
SELECT machine, counter FROM snowflake_get(lastval());
SELECT machine, counter FROM snowflake_get(nextval('snowflake_seq'));
SELECT machine, counter FROM snowflake_get(currval('snowflake_seq'));

-- Sequence relation exists, is unlogged and remains unlogged.
SELECT * FROM snowflake_seq;
ALTER SEQUENCE snowflake_seq SET LOGGED;
SELECT relpersistence FROM pg_class where relname = 'snowflake_seq';

ALTER SEQUENCE snowflake_seq RESTART;
SELECT * FROM snowflake_seq;

-- Identity column, where cache affects value.
SET default_sequence_access_method = 'snowflake';
CREATE TABLE snowflake_tab (a int GENERATED ALWAYS AS IDENTITY, b int);
INSERT INTO snowflake_tab VALUES (DEFAULT, generate_series(1, 10));
SELECT data.machine, data.counter
  FROM snowflake_tab, LATERAL snowflake_get(a) AS data;
DROP TABLE snowflake_tab;

DROP SEQUENCE snowflake_seq;
DROP EXTENSION snowflake;
