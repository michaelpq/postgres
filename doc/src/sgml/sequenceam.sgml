<!-- doc/src/sgml/sequenceam.sgml -->

<chapter id="sequenceam">
 <title>Sequence Access Method Interface Definition</title>

 <indexterm>
  <primary>Sequence Access Method</primary>
 </indexterm>
 <indexterm>
  <primary>sequenceam</primary>
  <secondary>Sequence Access Method</secondary>
 </indexterm>

 <para>
  This chapter explains the interface between the core
  <productname>PostgreSQL</productname> system and <firstterm>sequence access
  methods</firstterm>, which manage the operations around sequences . The core
  system knows little about these access methods beyond what is specified here,
  so it is possible to develop entirely new access method types by writing
  add-on code.
 </para>

 <para>
  Each sequence access method is described by a row in the
  <link linkend="catalog-pg-am"><structname>pg_am</structname></link> system
  catalog. The <structname>pg_am</structname> entry specifies a name and a
  <firstterm>handler function</firstterm> for the sequence access method.  These
  entries can be created and deleted using the
  <xref linkend="sql-create-access-method"/> and
  <xref linkend="sql-drop-access-method"/> SQL commands.
 </para>

 <para>
  A sequence access method handler function must be declared to accept a single
  argument of type <type>internal</type> and to return the pseudo-type
  <type>sequence_am_handler</type>.  The argument is a dummy value that simply
  serves to prevent handler functions from being called directly from SQL commands.

  The result of the function must be a pointer to a struct of type
  <structname>SequenceAmRoutine</structname>, which contains everything that the
  core code needs to know to make use of the sequence access method. The return
  value needs to be of server lifetime, which is typically achieved by
  defining it as a <literal>static const</literal> variable in global
  scope. The <structname>SequenceAmRoutine</structname> struct, also called the
  access method's <firstterm>API struct</firstterm>, defines the behavior of
  the access method using callbacks. These callbacks are pointers to plain C
  functions and are not visible or callable at the SQL level. All the
  callbacks and their behavior is defined in the
  <structname>SequenceAmRoutine</structname> structure (with comments inside
  the struct defining the requirements for callbacks). Most callbacks have
  wrapper functions, which are documented from the point of view of a user
  (rather than an implementor) of the sequence access method.  For details,
  please refer to the <ulink url="https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/include/access/sequenceam.h;hb=HEAD">
   <filename>src/include/access/sequenceam.h</filename></ulink> file.
 </para>

 <para>
  Currently, the way a sequence access method stores data is fairly
  unconstrained, and it is possible to use a predefined
  <link linkend="tableam">Table Access Method</link> to store sequence
  data.
 </para>

 <para>
  For crash safety, a sequence access method can use
  <link linkend="wal"><acronym>WAL</acronym></link>, or a custom
  implementation.
  If <acronym>WAL</acronym> is chosen, either
  <link linkend="generic-wal">Generic WAL Records</link> can be used, or a
  <link linkend="custom-rmgr">Custom WAL Resource Manager</link> can be
  implemented.
 </para>

 <para>
  Any developer of a new <literal>sequence access method</literal> can refer to
  the existing <literal>local</literal> implementation present in
  <filename>src/backend/access/sequence/local.c</filename> for details of
  its implementation.
 </para>
</chapter>
