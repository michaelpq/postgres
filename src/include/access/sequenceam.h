/*-------------------------------------------------------------------------
 *
 * sequenceam.h
 *	  POSTGRES sequence access method definitions.
 *
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/access/sequenceam.h
 *
 * NOTES
 *		See sequenceam.sgml for higher level documentation.
 *
 *-------------------------------------------------------------------------
 */
#ifndef SEQUENCEAM_H
#define SEQUENCEAM_H

#include "utils/rel.h"

#define DEFAULT_SEQUENCE_ACCESS_METHOD "seqlocal"

/* GUCs */
extern PGDLLIMPORT char *default_sequence_access_method;

/*
 * API struct for a sequence AM.  Note this must be allocated in a
 * server-lifetime manner, typically as a static const struct, which then gets
 * returned by FormData_pg_am.amhandler.
 *
 * In most cases it's not appropriate to call the callbacks directly, use the
 * sequence_* wrapper functions instead.
 *
 * GetSequenceAmRoutine() asserts that required callbacks are filled in,
 * remember to update when adding a callback.
 */
typedef struct SequenceAmRoutine
{
	/* this must be set to T_SequenceAmRoutine */
	NodeTag		type;

	/*
	 * Retrieve table access method used by a sequence to store its metadata.
	 */
	const char *(*get_table_am) (void);

	/*
	 * Initialize sequence after creating a sequence Relation in pg_class,
	 * setting up the sequence for use.  "last_value" and "is_called" are
	 * guessed from the options set for the sequence in CREATE SEQUENCE, based
	 * on the configuration of pg_sequence.
	 */
	void		(*init) (Relation rel, int64 last_value, bool is_called);

	/*
	 * Retrieve a result for nextval(), based on the options retrieved from
	 * the sequence's options in pg_sequence.  "last" is the last value
	 * calculated stored in the session's local cache, for lastval().
	 */
	int64		(*nextval) (Relation rel, int64 incby, int64 maxv,
							int64 minv, int64 cache, bool cycle,
							int64 *last);

	/*
	 * Callback to set the state of a sequence, based on the input arguments
	 * from setval().
	 */
	void		(*setval) (Relation rel, int64 next, bool iscalled);

	/*
	 * Reset a sequence to its initial value.  "reset_state", if set to true,
	 * means that the sequence parameters have changed, hence its internal
	 * state may need to be reset as well.  "startv" and "is_called" are
	 * values guessed from the configuration of the sequence, based on the
	 * contents of pg_sequence.
	 */
	void		(*reset) (Relation rel, int64 startv, bool is_called,
						  bool reset_state);

	/*
	 * Returns the current state of a sequence, returning data for
	 * pg_sequence_last_value() and related DDLs like ALTER SEQUENCE.
	 * "last_value" and "is_called" should be assigned to the values retrieved
	 * from the sequence Relation.
	 */
	void		(*get_state) (Relation rel, int64 *last_value, bool *is_called);

	/*
	 * Callback used when switching persistence of a sequence Relation, to
	 * reset the sequence based on its new persistence "newrelpersistence".
	 */
	void		(*change_persistence) (Relation rel, char newrelpersistence);

} SequenceAmRoutine;


/* ---------------------------------------------------------------------------
 * Wrapper functions for each callback.
 * ---------------------------------------------------------------------------
 */

/*
 * Returns the name of the table access method used by this sequence.
 */
static inline const char *
sequence_get_table_am(Relation rel)
{
	return rel->rd_sequenceam->get_table_am();
}

/*
 * Insert tuple data based on the information guessed from the contents
 * of pg_sequence.
 */
static inline void
sequence_init(Relation rel, int64 last_value, bool is_called)
{
	rel->rd_sequenceam->init(rel, last_value, is_called);
}

/*
 * Allocate a set of values for the given sequence.  "last" is the last value
 * allocated.  The result returned is the next value of the sequence computed.
 */
static inline int64
sequence_nextval(Relation rel, int64 incby, int64 maxv,
				 int64 minv, int64 cache, bool cycle,
				 int64 *last)
{
	return rel->rd_sequenceam->nextval(rel, incby, maxv, minv, cache,
									   cycle, last);
}

/*
 * Callback to set the state of a sequence, based on the input arguments
 * from setval().
 */
static inline void
sequence_setval(Relation rel, int64 next, bool iscalled)
{
	rel->rd_sequenceam->setval(rel, next, iscalled);
}

/*
 * Reset a sequence to its initial state.
 */
static inline void
sequence_reset(Relation rel, int64 startv, bool is_called,
			   bool reset_state)
{
	rel->rd_sequenceam->reset(rel, startv, is_called, reset_state);
}

/*
 * Retrieve sequence metadata.
 */
static inline void
sequence_get_state(Relation rel, int64 *last_value, bool *is_called)
{
	rel->rd_sequenceam->get_state(rel, last_value, is_called);
}

/*
 * Callback to change the persistence of a sequence Relation.
 */
static inline void
sequence_change_persistence(Relation rel, char newrelpersistence)
{
	rel->rd_sequenceam->change_persistence(rel, newrelpersistence);
}

/* ----------------------------------------------------------------------------
 * Functions in sequenceamapi.c
 * ----------------------------------------------------------------------------
 */

extern const SequenceAmRoutine *GetSequenceAmRoutine(Oid amhandler);
extern Oid	GetSequenceAmRoutineId(Oid amoid);

#endif							/* SEQUENCEAM_H */
