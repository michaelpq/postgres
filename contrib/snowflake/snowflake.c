/*-------------------------------------------------------------------------
 *
 * snowflake.c
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  contrib/snowflake/snowflake.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <sys/time.h>

#include "access/generic_xlog.h"
#include "access/sequenceam.h"
#include "access/xact.h"
#include "catalog/storage_xlog.h"
#include "commands/tablecmds.h"
#include "fmgr.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "nodes/makefuncs.h"
#include "storage/bufmgr.h"
#include "utils/guc.h"
#include "utils/numeric.h"

PG_MODULE_MAGIC;

/* "special area" of a snowflake's buffer page. */
#define SNOWFLAKE_MAGIC	0x01

typedef struct snowflake_magic
{
	uint32		magic;
} snowflake_magic;

/* -----------------------------------------------------------------------
 * Snowflake ID are 64-bit based, with the following structure:
 * - 41 bits for an epoch-based timestamp, in milli-seconds.
 * - 10 bits for a machine ID.
 * - 12 bits for a sequence counter.
 *
 * The timestamp can be cut to an offset.  The machine ID is controlled
 * by a superuser GUC.  Sequence properties apply to the sequence counter,
 * as the other two are environment-dependent.
 * -----------------------------------------------------------------------
 */

/*
 * Helper routines to convert a snowflake ID from/to an int64.
 */
#define SNOWFLAKE_COUNTER_MASK		0x0000000000000FFF	/* 12 bits */
#define SNOWFLAKE_COUNTER_SHIFT		0
#define SNOWFLAKE_MACHINE_ID_MASK	0x00000000000003FF	/* 10 bits */
#define SNOWFLAKE_MACHINE_ID_SHIFT	12	/* counter */
#define SNOWFLAKE_TIMESTAMP_MASK	0x000001FFFFFFFFFF	/* 41 bits */
#define SNOWFLAKE_TIMESTAMP_SHIFT	22	/* machine ID + counter sizes */

typedef struct snowflake_id
{
	uint32		count;
	uint32		machine;
	uint64		time_ms;
} snowflake_id;

#define snowflake_id_to_int64(id, raw) { \
	raw = (((id).count) & SNOWFLAKE_COUNTER_MASK) << SNOWFLAKE_COUNTER_SHIFT |	\
		(((id).machine) & SNOWFLAKE_MACHINE_ID_MASK) << SNOWFLAKE_MACHINE_ID_SHIFT | \
		(((id).time_ms) & SNOWFLAKE_TIMESTAMP_MASK) << SNOWFLAKE_TIMESTAMP_SHIFT;	\
}

#define int64_to_snowflake_id(raw, id) { \
	(id).count = ((raw) >> SNOWFLAKE_COUNTER_SHIFT) & SNOWFLAKE_COUNTER_MASK;		\
	(id).machine = ((raw) >> SNOWFLAKE_MACHINE_ID_SHIFT) & SNOWFLAKE_MACHINE_ID_MASK; \
	(id).time_ms = ((raw) >> SNOWFLAKE_TIMESTAMP_SHIFT) & SNOWFLAKE_TIMESTAMP_MASK;	\
}

/*
 * Format of tuples stored in heap table associated to snowflake sequence.
 */
typedef struct FormData_snowflake_data
{
	/* enough to cover 12 bits of the internal counter */
	int16	count;
	bool	is_called;
} FormData_snowflake_data;

typedef FormData_snowflake_data *Form_snowflake_data;

/*
 * Columns of a snowflake sequence relation.
 */
#define SNOWFLAKE_COL_COUNT		1
#define SNOWFLAKE_COL_CALLED	2

#define	SNOWFLAKE_COLS			2

/* GUC parameter */
static int snowflake_machine_id = 1;

PG_FUNCTION_INFO_V1(snowflake_sequenceam_handler);

/* -----------------------------------------------------------------------
 * Interfaces for relation manipulation.
 * -----------------------------------------------------------------------
 */

/*
 * Initialize snowflake relation's fork with some data.
 */
static void
fill_snowflake_fork(Relation rel, HeapTuple tuple, ForkNumber forkNum)
{
	Buffer		buf;
	Page		page;
	snowflake_magic *sm;
	OffsetNumber offnum;
	GenericXLogState *state = NULL;

	/* Initialize first page of relation with special magic number */

	buf = ExtendBufferedRel(BMR_REL(rel), forkNum, NULL,
							EB_LOCK_FIRST | EB_SKIP_EXTENSION_LOCK);
	Assert(BufferGetBlockNumber(buf) == 0);

	page = BufferGetPage(buf);

	PageInit(page, BufferGetPageSize(buf), sizeof(snowflake_magic));
	sm = (snowflake_magic *) PageGetSpecialPointer(page);
	sm->magic = SNOWFLAKE_MAGIC;

	/* Now insert sequence tuple */

	/*
	 * Since VACUUM does not process sequences, we have to force the tuple to
	 * have xmin = FrozenTransactionId now.  Otherwise it would become
	 * invisible to SELECTs after 2G transactions.  It is okay to do this
	 * because if the current transaction aborts, no other xact will ever
	 * examine the sequence tuple anyway.
	 */
	HeapTupleHeaderSetXmin(tuple->t_data, FrozenTransactionId);
	HeapTupleHeaderSetXminFrozen(tuple->t_data);
	HeapTupleHeaderSetCmin(tuple->t_data, FirstCommandId);
	HeapTupleHeaderSetXmax(tuple->t_data, InvalidTransactionId);
	tuple->t_data->t_infomask |= HEAP_XMAX_INVALID;
	ItemPointerSet(&tuple->t_data->t_ctid, 0, FirstOffsetNumber);

	/*
	 * Initialize before entering in the critical section, as this does
	 * allocations.
	 */
	if (forkNum == INIT_FORKNUM)
		state = GenericXLogStart(rel);

	START_CRIT_SECTION();

	MarkBufferDirty(buf);

	offnum = PageAddItem(page, (Item) tuple->t_data, tuple->t_len,
						 InvalidOffsetNumber, false, false);
	if (offnum != FirstOffsetNumber)
		elog(ERROR, "failed to add sequence tuple to page");

	/*
	 * Init forks have to be logged.  These go through generic WAL records
	 * for simplicity's sake to save from the need of a custom RMGR.
	 */
	if (forkNum == INIT_FORKNUM)
	{
		(void) GenericXLogRegisterBuffer(state, buf, GENERIC_XLOG_FULL_IMAGE);
		GenericXLogFinish(state);
	}

	END_CRIT_SECTION();

	UnlockReleaseBuffer(buf);
}

/*
 * Initialize snowflake relation.
 *
 * This needs to handle both the initial and main forks.
 */
static void
fill_snowflake(Relation rel, HeapTuple tuple)
{
	SMgrRelation srel;

	Assert(rel->rd_rel->relpersistence == RELPERSISTENCE_UNLOGGED);

	fill_snowflake_fork(rel, tuple, MAIN_FORKNUM);

	/* init fork */
	srel = smgropen(rel->rd_locator, INVALID_PROC_NUMBER);
	smgrcreate(srel, INIT_FORKNUM, false);
	log_smgrcreate(&rel->rd_locator, INIT_FORKNUM);
	fill_snowflake_fork(rel, tuple, INIT_FORKNUM);
	FlushRelationBuffers(rel);
	smgrclose(srel);
}

/*
 * Read the current state of a snowflake sequence
 *
 * Given an opened sequence relation, lock the page buffer and find the tuple.
 *
 * *buf receives the reference to the pinned-and-ex-locked buffer.
 * *seqdatatuple receives the reference to the sequence tuple proper.
 *
 * Returns value points to the data payload of the tuple.
 */
static Form_snowflake_data
read_snowflake(Relation rel, Buffer *buf, HeapTuple seqdatatuple)
{
	Page		page;
	ItemId		lp;
	snowflake_magic *sm;
	Form_snowflake_data	seq;

	*buf = ReadBuffer(rel, 0);
	LockBuffer(*buf, BUFFER_LOCK_EXCLUSIVE);

	page = BufferGetPage(*buf);
	sm = (snowflake_magic *) PageGetSpecialPointer(page);

	if (sm->magic != SNOWFLAKE_MAGIC)
		elog(ERROR, "bad magic number in sequence \"%s\": %08X",
			 RelationGetRelationName(rel), sm->magic);

	lp = PageGetItemId(page, FirstOffsetNumber);
	Assert(ItemIdIsNormal(lp));

	/* Note we currently only bother to set these two fields of *seqdatatuple */
	seqdatatuple->t_data = (HeapTupleHeader) PageGetItem(page, lp);
	seqdatatuple->t_len = ItemIdGetLength(lp);

	/*
	 * Previous releases of Postgres neglected to prevent SELECT FOR UPDATE on
	 * a sequence, which would leave a non-frozen XID in the sequence tuple's
	 * xmax, which eventually leads to clog access failures or worse. If we
	 * see this has happened, clean up after it.  We treat this like a hint
	 * bit update, ie, don't bother to WAL-log it, since we can certainly do
	 * this again if the update gets lost.
	 */
	Assert(!(seqdatatuple->t_data->t_infomask & HEAP_XMAX_IS_MULTI));
	if (HeapTupleHeaderGetRawXmax(seqdatatuple->t_data) != InvalidTransactionId)
	{
		HeapTupleHeaderSetXmax(seqdatatuple->t_data, InvalidTransactionId);
		seqdatatuple->t_data->t_infomask &= ~HEAP_XMAX_COMMITTED;
		seqdatatuple->t_data->t_infomask |= HEAP_XMAX_INVALID;
		MarkBufferDirtyHint(*buf, true);
	}

	seq = (Form_snowflake_data) GETSTRUCT(seqdatatuple);

	return seq;
}


/* ------------------------------------------------------------------------
 * Callbacks for the snowflake sequence access method.
 * ------------------------------------------------------------------------
 */

/*
 * Return the table access method used by this sequence.
 *
 * This is just an on-memory sequence, so anything is fine.
 */
static const char *
snowflake_sequenceam_get_table_am(void)
{
	return "heap";
}

/*
 * snowflake_sequenceam_init
 *
 * Initialize relation of a snowflake sequence.  This stores the sequence
 * counter in an unlogged relation as timestamps ensure value unicity.
 */
static void
snowflake_sequenceam_init(Relation rel, int64 last_value, bool is_called)
{
	Datum	values[SNOWFLAKE_COLS];
	bool	nulls[SNOWFLAKE_COLS];
	int16	counter;
	TupleDesc	tupdesc;
	HeapTuple	tuple;
	List	   *elts = NIL;
	ListCell   *lc;
	ColumnDef  *coldef = NULL;
	AlterTableCmd *atcmd;
	List		  *atcmds = NIL;

	/* Adjust last_value, depending on the defaults given */
	counter = ((int16) last_value) & SNOWFLAKE_COUNTER_MASK;

	/*
	 * Create unlogged relation with its attributes.
	 */
	coldef = makeColumnDef("count", INT2OID, -1, InvalidOid);
	coldef->is_not_null = true;
	elts = lappend(elts, coldef);
	coldef = makeColumnDef("is_called", BOOLOID, -1, InvalidOid);
	coldef->is_not_null = true;
	elts = lappend(elts, coldef);

	foreach(lc, elts)
	{
		atcmd = makeNode(AlterTableCmd);
		atcmd->subtype = AT_AddColumnToSequence;
		atcmd->def = (Node *) lfirst(lc);
		atcmds = lappend(atcmds, atcmd);
	}

	/*
	 * No recursion needed.  Note that EventTriggerAlterTableStart() should
	 * have been called.
	 */
	AlterTableInternal(RelationGetRelid(rel), atcmds, false);
	CommandCounterIncrement();

	/*
	 * Switch the relation to be unlogged.  This forces a rewrite, but
	 * the relation is empty so that's OK.
	 */
	RelationSetNewRelfilenumber(rel, RELPERSISTENCE_UNLOGGED);

	/* And insert its first tuple */
	values[0] = Int16GetDatum(counter);
	nulls[0] = false;
	values[1] = BoolGetDatum(is_called);
	nulls[1] = false;

	tupdesc = RelationGetDescr(rel);
	tuple = heap_form_tuple(tupdesc, values, nulls);
	fill_snowflake(rel, tuple);
}

/*
 * snowflake_sequenceam_nextval
 *
 * Return the next value for a snowflake sequence.
 */
static int64
snowflake_sequenceam_nextval(Relation rel, int64 incby, int64 maxv,
							 int64 minv, int64 cache, bool cycle,
							 int64 *last)
{
	Buffer		buf;
	Form_snowflake_data seq;
	HeapTupleData seqdatatuple;
	int64	result = 0;
	snowflake_id	id = {0};
	struct timeval tp;

	/* lock page buffer and read tuple */
	seq = read_snowflake(rel, &buf, &seqdatatuple);

	/*
	 * The logic here is quite simple, increment the counter until its
	 * threshold is reached and get back to the start.  If the threshold
	 * is reached, wait 1ms to ensure a unique timestamp.  There is no
	 * need to do a retry as the buffer is already locked.
	 */
	id.count = seq->count;
	id.count++;

	if (id.count > (PG_INT16_MAX & SNOWFLAKE_COUNTER_MASK))
	{
		/*
		 * Threshold reached, so wait a bit for force clock to a new
		 * timestamp.
		 */
		id.count = 1;
		pg_usleep(1000L);	/* 1ms */
	}

	/* Compute timestamp, with buffer locked */
	gettimeofday(&tp, NULL);
	id.time_ms = (uint64) tp.tv_sec * 1000 +
		tp.tv_usec / 1000;

	/* Machine ID */
	id.machine = snowflake_machine_id;

	/* ready to change the on-disk (or really, in-buffer) tuple */
	START_CRIT_SECTION();
	seq->count = id.count;
	seq->is_called = true;
	END_CRIT_SECTION();

	UnlockReleaseBuffer(buf);

	/* Store the last value computed for lastval() */
	snowflake_id_to_int64(id, result);
	*last = result;
	return result;
}

/*
 * snowflake_sequenceam_setval
 *
 * Set the sequence value, manipulating only the sequence counter.
 */
static void
snowflake_sequenceam_setval(Relation rel, int64 next, bool iscalled)
{
	Buffer		buf;
	HeapTupleData seqdatatuple;
	Form_snowflake_data seq;

	/* lock page buffer and read tuple */
	seq = read_snowflake(rel, &buf, &seqdatatuple);

	/* Change the in-buffer tuple */
	START_CRIT_SECTION();
	seq->count = (next & SNOWFLAKE_COUNTER_MASK);
	seq->is_called = iscalled;
	MarkBufferDirty(buf);
	END_CRIT_SECTION();

	UnlockReleaseBuffer(buf);
}

/*
 * snowflake_sequenceam_get_state
 *
 * Return the last sequence counter value.
 */
static void
snowflake_sequenceam_get_state(Relation rel, int64 *last_value, bool *is_called)
{
	Buffer		buf;
	HeapTupleData seqdatatuple;
	Form_snowflake_data seq;

	seq = read_snowflake(rel, &buf, &seqdatatuple);
	*last_value = seq->count;
	*is_called = seq->is_called;
	UnlockReleaseBuffer(buf);
}

/*
 * snowflake_sequenceam_reset
 *
 * Reset the sequence, coming down to resetting its counter.
 */
static void
snowflake_sequenceam_reset(Relation rel, int64 startv, bool is_called,
						   bool reset_state)
{
	HeapTupleData seqdatatuple;
	HeapTuple	tuple;
	Form_snowflake_data seq;
	Buffer		buf;

	/* lock buffer page and read tuple */
	(void) read_snowflake(rel, &buf, &seqdatatuple);

	/* copy the existing tuple */
	tuple = heap_copytuple(&seqdatatuple);

	/* Now we're done with the old page */
	UnlockReleaseBuffer(buf);

	/*
	 * Modify the copied tuple to execute the restart (compare the RESTART
	 * action in AlterSequence)
	 */
	seq = (Form_snowflake_data) GETSTRUCT(tuple);
	seq->count = (startv & SNOWFLAKE_COUNTER_MASK);
	seq->is_called = is_called;

	/* create new storage */
	RelationSetNewRelfilenumber(rel, rel->rd_rel->relpersistence);

	/* insert the modified tuple into the page */
	fill_snowflake(rel, tuple);
}

/*
 * snowflake_sequenceam_change_persistence
 *
 * There is nothing to do here, the underneath relation has to remain
 * unlogged and is set as such when creating the sequence.
 */
static void
snowflake_sequenceam_change_persistence(Relation rel, char newrelpersistence)
{
	/* Nothing to do here */
}

/* ------------------------------------------------------------------------
 * Definition of the snowflake sequence access method.
 * ------------------------------------------------------------------------
 */

static const SequenceAmRoutine snowflake_sequenceam_methods = {
	.type = T_SequenceAmRoutine,
	.get_table_am = snowflake_sequenceam_get_table_am,
	.init = snowflake_sequenceam_init,
	.nextval = snowflake_sequenceam_nextval,
	.setval = snowflake_sequenceam_setval,
	.get_state = snowflake_sequenceam_get_state,
	.reset = snowflake_sequenceam_reset,
	.change_persistence = snowflake_sequenceam_change_persistence
};

Datum
snowflake_sequenceam_handler(PG_FUNCTION_ARGS)
{
	PG_RETURN_POINTER(&snowflake_sequenceam_methods);
}

/* Utility functions */

/*
 * snowflake_get
 *
 * Return a tuple worth of snowflake ID data, in a readable shape.
 */
PG_FUNCTION_INFO_V1(snowflake_get);
Datum
snowflake_get(PG_FUNCTION_ARGS)
{
#define	SNOWFLAKE_GET_COLS	3
	int64	raw = PG_GETARG_INT64(0);
	Datum  *values;
	bool   *nulls;
	TupleDesc		tupdesc;
	snowflake_id	id;

	/* determine result type */
	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
		elog(ERROR, "return type must be a row type");

	int64_to_snowflake_id(raw, id);

	nulls = palloc0(sizeof(bool) * tupdesc->natts);
	values = palloc0(sizeof(Datum) * tupdesc->natts);

	values[0] = Int64GetDatum(id.time_ms);
	values[1] = Int32GetDatum(id.machine);
	values[2] = Int32GetDatum(id.count);

	/* Returns the record as Datum */
	PG_RETURN_DATUM(HeapTupleGetDatum(heap_form_tuple(tupdesc, values, nulls)));
}

/*
 * Entry point when loading extension.
 */
void
_PG_init(void)
{
	DefineCustomIntVariable("snowflake.machine_id",
							"Machine ID to use with snowflake sequence.",
							"Default value is 1.",
							&snowflake_machine_id,
							1, 0, 1023, /* 10 bits as max */
							PGC_SUSET,
							0, NULL, NULL, NULL);
}
