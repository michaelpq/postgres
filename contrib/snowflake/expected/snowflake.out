CREATE EXTENSION snowflake;
CREATE SEQUENCE snowflake_seq USING snowflake;
SET snowflake.machine_id = 2000; -- error
ERROR:  2000 is outside the valid range for parameter "snowflake.machine_id" (0 .. 1023)
SET snowflake.machine_id = 4; -- ok
SELECT machine, counter FROM snowflake_get(nextval('snowflake_seq'));
 machine | counter 
---------+---------
       4 |       2
(1 row)

SELECT machine, counter FROM snowflake_get(lastval());
 machine | counter 
---------+---------
       4 |       2
(1 row)

SELECT machine, counter FROM snowflake_get(nextval('snowflake_seq'));
 machine | counter 
---------+---------
       4 |       3
(1 row)

SELECT machine, counter FROM snowflake_get(currval('snowflake_seq'));
 machine | counter 
---------+---------
       4 |       3
(1 row)

-- Sequence relation exists, is unlogged and remains unlogged.
SELECT * FROM snowflake_seq;
 count | is_called 
-------+-----------
     3 | t
(1 row)

ALTER SEQUENCE snowflake_seq SET LOGGED;
SELECT relpersistence FROM pg_class where relname = 'snowflake_seq';
 relpersistence 
----------------
 u
(1 row)

ALTER SEQUENCE snowflake_seq RESTART;
SELECT * FROM snowflake_seq;
 count | is_called 
-------+-----------
     1 | f
(1 row)

-- Identity column, where cache affects value.
SET default_sequence_access_method = 'snowflake';
CREATE TABLE snowflake_tab (a int GENERATED ALWAYS AS IDENTITY, b int);
INSERT INTO snowflake_tab VALUES (DEFAULT, generate_series(1, 10));
SELECT data.machine, data.counter
  FROM snowflake_tab, LATERAL snowflake_get(a) AS data;
 machine | counter 
---------+---------
       4 |       2
       4 |       3
       4 |       4
       4 |       5
       4 |       6
       4 |       7
       4 |       8
       4 |       9
       4 |      10
       4 |      11
(10 rows)

DROP TABLE snowflake_tab;
DROP SEQUENCE snowflake_seq;
DROP EXTENSION snowflake;
